hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/AbstractContainerReportHandler.java

```diff
@@ -352,26 +352,32 @@ private boolean updateContainerState(final DatanodeDetails datanode,
        */
       break;
     case DELETING:
+    case DELETED:
       /*
-      HDDS-11136: If a DELETING container has a non-empty CLOSED replica, the container should be moved back to CLOSED
-      state.
+       * HDDS-11136: If a DELETING container has a non-empty CLOSED replica, the container should be moved back to
+       * CLOSED state.
+       *
+       * HDDS-12421: If a DELETING or DELETED container has a non-empty replica, the container should also be moved
+       * back to CLOSED state.
        */
-      boolean replicaStateAllowed = replica.getState() == State.CLOSED;
-      boolean replicaNotEmpty = replica.hasIsEmpty() && !replica.getIsEmpty();
-      if (replicaStateAllowed && replicaNotEmpty) {
-        logger.info("Moving DELETING container {} to CLOSED state, datanode {} reported replica with state={}, " +
-            "isEmpty={} and keyCount={}.", containerId, datanode.getHostName(), replica.getState(), false,
-            replica.getKeyCount());
-        containerManager.transitionDeletingToClosedState(containerId);
+      boolean replicaIsEmpty = replica.hasIsEmpty() && replica.getIsEmpty();
+      // If container is in DELETED state and the reported replica is empty, delete the empty replica.
+      // We should also do this for DELETING containers and currently DeletingContainerHandler does that
+      if (container.getState() == HddsProtos.LifeCycleState.DELETED && replicaIsEmpty) {
+        // should we send a non-force delete instead?
```

Comment:
Yes we should use non-force delete in all cases except over-replication. Good catch I missed that this method was using force internally. We can add `force` as a parameter to `deleteReplica`, it only has two callers.

Reply:
In a future change we could consider a refactoring where force is never given as a parameter. All methods would be of the form `sendDeleteCommand` or `sendForceDeleteCommand`. `DeleteContainerCommand` could have two static factories to build force and non-force variants with private constructors. This would make it clearer when reading the code which paths are doing the "dangerous" force delete operation.

Reply:
Good catch @siddhantsangwan @errose28 . Done.

> All methods would be of the form sendDeleteCommand or sendForceDeleteCommand. DeleteContainerCommand could have two static factory methods to build force and non-force variants with private constructors. This would make it clearer when reading the code which paths are doing the "dangerous" force delete operation.

Good idea. But I won't do the refactoring here. This will be in another PR.

Reply:
Filed HDDS-12447